gpu/src/cuda/shaders/u128.cuh

```diff
@@ -0,0 +1,135 @@
+    return u128(high - rhs.high - ((low - rhs.low) > low), low - rhs.low);
+  }
+
+  __device__ constexpr u128 operator-=(const u128 rhs) {
+    *this = *this - rhs;
+    return *this;
+  }
+
+  __device__ constexpr bool operator==(const u128 rhs) const {
+    return high == rhs.high && low == rhs.low;
+  }
+
+  __device__ constexpr bool operator!=(const u128 rhs) const {
+    return !(*this == rhs);
+  }
+
+  __device__ constexpr bool operator<(const u128 rhs) const {
+    return ((high == rhs.high) && (low < rhs.low)) || (high < rhs.high);
+  }
+
+  __device__ constexpr u128 operator&(const u128 rhs) const {
+    return u128(high & rhs.high, low & rhs.low);
+  }
+
+  __device__ constexpr u128 operator|(const u128 rhs) const {
+    return u128(high | rhs.high, low | rhs.low);
+  }
+
+  __device__ constexpr bool operator>(const u128 rhs) const {
+    return ((high == rhs.high) && (low > rhs.low)) || (high > rhs.high);
+  }
+
+  __device__ constexpr bool operator>=(const u128 rhs) const {
+    return !(*this < rhs);
+  }
+
+  __device__ constexpr bool operator<=(const u128 rhs) const {
+    return !(*this > rhs);
+  }
+
+  __device__ constexpr inline u128 operator>>(unsigned shift) const {
+    // TODO: reduce branch conditions
+    if (shift >= 128) {
+      return u128(0);
+    } else if (shift == 64) {
+      return u128(0, high);
+    } else if (shift == 0) {
+      return *this;
+    } else if (shift < 64) {
+      return u128(high >> shift, (high << (64 - shift)) | (low >> shift));
+    } else if ((128 > shift) && (shift > 64)) {
+      return u128(0, (high >> (shift - 64)));
+    } else {
+      return u128(0);
+    }
+  }
+
+  __device__ constexpr inline u128 operator<<(unsigned shift) const {
+    // TODO: reduce branch conditions
+    if (shift >= 128) {
+      return u128(0);
+    } else if (shift == 64) {
+      return u128(low, 0);
+    } else if (shift == 0) {
+      return *this;
+    } else if (shift < 64) {
+      return u128((high << shift) | (low >> (64 - shift)), low << shift);
+    } else if ((128 > shift) && (shift > 64)) {
+      return u128((low >> (shift - 64)), 0);
+    } else {
+      return u128(0);
+    }
+  }
+
+  __device__ constexpr u128 operator>>=(unsigned rhs) {
+    *this = *this >> rhs;
+    return *this;
+  }
+
+  __device__ u128 operator*(const bool rhs) const {
+    return u128(high * rhs, low * rhs);
+  }
+
+  __device__ u128 operator*(const u128 rhs) const {
+    unsigned long t_low_high = __umul64hi(low, rhs.high);
+    unsigned long t_high = __umul64hi(low, rhs.low);
+    unsigned long t_high_low = __umul64hi(high, rhs.low);
+    unsigned long t_low = low * rhs.low;
+    return u128(t_low_high + t_high_low + t_high, t_low);
+  }
+
+  __device__ u128 operator*=(const u128 rhs) {
+    *this = *this * rhs;
+    return *this;
+  }
+
+  unsigned long high;
+  unsigned long low;
+
+  // TODO: Could get better performance with  smaller limb size
+  // Not sure what word size is for M1 GPU
```

Comment:
Is this relevant with CUDA? Doesn't it only support Metal?

Reply:
No, most code was ported from the metal implementation and this comment wasn't updated. Changed for a reminder about a possible optimization in changing the limb size.